narwhals/dtypes.py

```diff
@@ -460,35 +461,62 @@ class Categorical(DType):
     """
 
 
+class _DelayedCategories:
+    """Store callable which produces tupleified version of Enum's categories.
+
+    !!! warning
+        This class is not meant to be instantiated directly and exists
+        for internal usage only.
+    """
+
+    def __init__(self, get_categories: Callable[[], tuple[str, ...]]) -> None:
+        self.get_categories = get_categories
+
+
 class Enum(DType):
     """A fixed categorical encoding of a unique set of strings.
 
-    Polars has an Enum data type, while pandas and PyArrow do not.
+    Polars has an Enum data type. In pandas, ordered categories get mapped
+    to Enum. PyArrow has no Enum equivalent.
 
     Examples:
        >>> import narwhals as nw
        >>> nw.Enum(["beluga", "narwhal", "orca"])
        Enum(categories=['beluga', 'narwhal', 'orca'])
     """
 
-    categories: Sequence[str]
+    def __init__(
+        self, categories: Iterable[str] | type[enum.Enum] | _DelayedCategories
+    ) -> None:
```

Comment:
Dependent on (https://github.com/narwhals-dev/narwhals/pull/2414#discussion_r2054049105)

But checks out after making that change 🙂
```suggestion
    def __init__(self, categories: Iterable[str] | type[enum.Enum]) -> None:
```